### DISCUSSION TITLE
How to specify the order of back press handlers?

### DISCUSSION THREAD

**Discussion by evjava** [arkivanov/Decompose#36] [Q&A] [answered]
I want to specify the order of handlers like follows:
1. close drawer (_if open_)
2. close search (_if expanded_)
3. handle backpress by component (_if can_)
4. router.pop
4. handle double-press (_first press for toast "press twice to exit", second quick press for exit_)

I can implement all these handlers, but I don't understand, how to tangle them together in __correct order__.

Is there some API to do it? [As I see](https://github.com/arkivanov/Essenty/blob/767259397cf8c8ccea8b6f82b063cad9d8214414/back-pressed/src/commonMain/kotlin/com/arkivanov/essenty/backpressed/DefaultBackPressedDispatcher.kt#L21), the order isn't specified, moreover, router calls pop [after](https://github.com/arkivanov/Decompose/blob/3c2a1f3bfd8b7d1ebbac3286ddd4da867e387e35/decompose/src/commonMain/kotlin/com/arkivanov/decompose/router/RouterImpl.kt#L46) other handlers.

_Sub question_: is it possible to access `router.backstackSize` from outside? Probably it can help me.

**Comment by arkivanov** [comment, marked as answer]
Hello and thanks for the question!

>  As I see, the order isn't specified

The `BackPresessedHandler` is part of the [Essenty](https://github.com/arkivanov/Essenty) library. And the order is [specified in the README file](https://github.com/arkivanov/Essenty#content-3). All registered callbacks are called in reverse order. This reflects the behaviour of the the [AndroidX OnBackPressedDispatcher](https://developer.android.com/reference/androidx/activity/OnBackPressedDispatcher).

> I can implement all these handlers, but I don't understand, how to tangle them together in correct order.

Based on what you described, I suggest the following implementation. First, you can avoid passing the `handleBackButton` flag to the `ComponentContext.router(...)` method. This will allow manual back button handling. Next, perhaps you can register just one callback to the `BackButtonHandler` and do checks and actions there. Return `true` to intercept the back button handling, or `false` to allow other registered callbacks.

> is it possible to access `router.backstackSize` from outside?

Yes it is possible. Will `router.state.value.backstack.size` work for you?

**Reply by arkivanov** [reply]
Or if the drawer/search logic is in a child component, then the following approach can help.

In a child component:

```kotlin
backPressedHandler.register {
    when {
        drawer.isOpen -> {
            drawer.close()
            true
        }
        search.isOpen -> {
            search.close()
            true
        }
        else -> {
            // Handle the back button
            true | false // Return true to intercept the event or false to allow other callbacks
        }
    }
}
```

In the root:

```kotlin
init {
    backPressedHandler.register {
        // Called if none of the children intercepted the event and the router has nothing to pop
        // Check the timeout and show the toast if needed
        true | false  // Return true if the toast was shown, false otherwise
    }
}

private val router = router(handleBackButton = true, ...)
```

**Reply by evjava** [reply]
`> Next, perhaps you can register just one callback to the BackButtonHandler and do checks and actions there.`

Yep, it works, thanks!

`> Will router.state.value.backstack.size work for you?`

Yes. I didn't notice it.. 🤔
